medmodels/treatment_effect_estimation/treatment_effect.py

```diff
@@ -0,0 +1,512 @@
+            edges_one_direction = self.medrecord.edges_connecting(treatment, node_id)
+            edges_other_direction = self.medrecord.edges_connecting(node_id, treatment)
+
+            edges = edges_one_direction + edges_other_direction
+            # If the node does not have the treatment, continue
+            if not edges:
+                continue
+
+            # If the node has the treatment, check if it has the time attribute
+            edge_values = self.medrecord.edge[edges].values()
+            assert all(
+                self.time_attribute in edge_attribute for edge_attribute in edge_values
+            ), "Time attribute not found in the edge attributes"
+
+            # Find the minimum time of the treatments
+            edge_times = [
+                pd.to_datetime(str(edge_attribute[self.time_attribute]))
+                for edge_attribute in edge_values
+            ]
+            if edge_times:
+                min_time = min(edge_times)
+                time_treat = min(min_time, time_treat)
+
+        assert (
+            time_treat != pd.Timestamp.max
+        ), f"No treatment found for node {node_id} in this MedRecord"
+
+        return time_treat
+
+    def find_controls(
+        self,
+        criteria_filter: Optional[NodeOperation] = None,
+    ) -> Tuple[Set[NodeIndex], Set[NodeIndex]]:
+        """
+        Identifies control groups among patients who did not undergo the specified
+        treatments. Control groups are divided into those who had the outcome
+        (control_true) and those who did not (control_false), based on the presence of
+        specified outcome codes.
+
+        Args:
+            criteria_filter (NodeOperation): Criteria to filter patients for the
+                control groups. Defaults to an empty list, meaning no additional
+                filtering criteria.
+
+        Returns:
+            Tuple[Set[NodeIndex], Set[NodeIndex]]: Two sets representing the IDs of
+                control patients. The first set includes patients who experienced the
+                specified outcomes (control_true), and the second set includes those who
+                did not (control_false).
+
+        This method facilitates the separation of patients into relevant control groups
+        for further analysis of treatment effects, ensuring only those not undergoing
+        the treatment are considered for control.
+        """
+        if not criteria_filter:
+            # If no criteria are given, use all the patients
+            criteria_set = set(self.medrecord.group(self.patients_group))
+        else:
+            nodes_to_add = self.medrecord.select_nodes(
+                node().index().is_in(self.medrecord.group(self.patients_group))
+                & criteria_filter
+            )
+            criteria_set = set(nodes_to_add)
+
+        control_all = criteria_set - self.treatment_all
+        assert (
+            control_all != set()
```

Comment:
Does this work correctly?

Checking for size == 0 would be more verbose IMO

Reply:
It does work, as it is checked for in the test. I will change it to len()>0 for better readability.